### ISSUE TITLE
SimpliSafe Camera Integration

### ISSUE THREAD

**Issue by tbrock47** [bachya/simplisafe-python#85]
So the "SimpliCam" as they call it is a combination camera and motion sensor.
The camera itself is joined to your local network wifi during setup. I am unsure if the included sensor is also communicating via the local network or their standard communication protocol.

You stated you don't have a camera of your own, so let me know how I can help.

**Comment by bachya** [comment]
Thanks for offering to help! I need to understand how the SimpliSafe web app communicates with the camera. Here's my suggestion:

1. Open Google Chrome.
2. Navigate to https://webapp.simplisafe.com and login.
3. Open the Chrome Developer Tools (`View >> Developer >> Developer Tools`).
4. Browse around various camera "operations" – open the camera page, view the camera, stop viewing the camera, go elsewhere, etc.
5. When done, go back to the Developer Tools and click the `Export HAR` link (see below). Don't post that here; send it to me in a DM on the HASS Community Forum.

![Screen Shot 2020-01-28 at 11 33 14 AM](https://user-images.githubusercontent.com/47216/73293929-0580a800-41c2-11ea-8806-0ef7e9a1d742.png)

**Comment by bachya** [comment]
Thanks for the payload, @tbrock47. Here's what I've learned:

Once you log into the web app and access the camera page, a URL like this is request:

```
https://media.simplisafe.com/v1/<SOME_ID>/flv?x=472&audioEncoding=AAC
```

I assume that this is the stream URL to access, although when I attempted to access yours, I got `Unauthorized`. I'm assuming this is either because (a) the access token is shorted lived or (b) I got the headers wrong.

I'll continue to investigate.

**Comment by KTibow** [comment]
Also, how does the operation of viewing previous motion events work?

**Comment by tbrock47** [comment]
> Also, how does the operation of viewing previous motion events work?

The motion events aren't logged on the Cameras page. They show up as a line item under the Timeline page as "Camera Motion Detected". The Cameras page is just a tile layout of all your cameras which allows you to click on one to get a live view.

I'm honestly more interested in just getting a live stream over anything else. The reset would be a bonus for a later date IMO.

**Comment by KTibow** [comment]
Got it. But once you get the livestream going, when you add it to Home Assistant, you probably won’t want to have live view always on. (Maybe take a “snapshot” every hour.)

**Comment by ph1ash** [comment]
Hey all - been digging into this as well. So the `<SOME_ID>` (which is a unique camera ID) mentioned above comes through during the `https://api.simplisafe.com/v1/users/<user_id>/subscriptions` GET request. In the JSON response, you'll find your camera ID in `subscriptions-><first subscription entry>->location->system->cameras-><first entry>->uuid` . This is then used to point to the link mentioned above; `https://media.simplisafe.com/v1/<Camera UUID>/flv?x=472&audioEncoding=AAC`. I'm having trouble with the `get_systems` call at the moment, but will try to resolve that soon and then try fetching the uuid through the `get_systems` API (might already be coming through but I can't tell unfortunately).

**Comment by bachya** [comment]
Thanks @ph1ash! Let me know what I can to to assist you.

**Comment by archerne** [comment]
Out of curiosity, the SimpliSafe camera is available to view on the SimpliSafe app without any subscription, are you looking into a way to be able to view the camera without pulling the subscription settings?

**Comment by jaredswarts55** [comment]
I don't really know what I am doing in python but I was able to stream the flv to a file and watch it with vlc using this code:
```

import asyncio

from aiohttp import ClientSession
from simplipy import *


async def get_subscriptions(user_id: str, access_token: str, client: ClientSession):
    url = 'https://api.simplisafe.com/v1/users/{}/subscriptions'.format(user_id)
    print("Url To Send Call To: {}".format(url))
    async with client.get(url, headers={'Authorization': 'Bearer {}'.format(access_token)}) as resp:
        return await resp.json()

async def get_camera_feed(camera_id: str, access_token: str, client: ClientSession):
    url = 'https://media.simplisafe.com/v1/{}/flv?x=472&audioEncoding=AAC'.format(camera_id)
    print("Url To Send Call To: {}".format(url))
    with open('testing.flv', 'wb') as file:
        async with client.get(url, headers={'Authorization': 'Bearer {}'.format(access_token)}) as resp:
            async for data in resp.content.iter_chunked(1024):
                print(data);
                file.write(data)

async def main() -> None:
    """Create the aiohttp session and run."""
    async with ClientSession() as session:
        simplisafe = await API.login_via_credentials(
            "YOUR_EMAIL_HERE", "YOUR_PASSWORD_HERE", session=session
        )

        systems = await simplisafe.get_systems();
        for s in systems:
            system = systems[s]

        subscriptionsBody = await get_subscriptions(simplisafe.user_id, simplisafe.access_token, session)
        cameraId = subscriptionsBody["subscriptions"][0]["location"]["system"]["cameras"][0]["uuid"]
        print("Camera Id: {}".format(cameraId))
        print("Subscriptions: {}".format(subscriptionsBody))
        await get_camera_feed(cameraId, simplisafe.access_token, session)


asyncio.run(main())
```

**Comment by KTibow** [comment]
@jaredswarts55 what do you mean "stream"? Did you just press Ctrl+C to stop it when you were done? I'd be interested if there was a way to get **past** camera recordings too.

**Comment by jaredswarts55** [comment]
@KTibow yeah, the request was open and streaming to the file, increasing its file size, I had to hit Ctrl+C to stop it.

**Comment by KTibow** [comment]
@jaredswarts55 so at the end of it did VLC complain about invalidness, or was it okay since you chunked them into 1024-byte parts?

**Comment by nikonratm** [comment]
Hey guys, I work on a node implementation of the SS api for Apple HomeKit [nzapponi/homebridge-simplisafe3](https://github.com/nzapponi/homebridge-simplisafe3), we've had camera integration for a while though obviously its different we pipe through ffmpeg to an rstp stream (which is what Apple requires).

Unfortunately it seems theres no way to locally tap into the cameras as you've noted but you guys have the right URL FWIW, theres also an mjpeg stream available at:
`https://media.simplisafe.com/v1/{cameraID}/mjpg?x=${width}&fr=1`, and finally wanted to mention if you leave off the `audioEncoding=AAC` you will get the h.264 stream with speex audio instead of aac.

Good luck! I will be following along in case I can learn something for my project =)

**Comment by jaredswarts55** [comment]
@KTibow vlc is pretty good about continuously reading from the disk and doing its best to watch the video stream, when I ended it, vlc stopped like the video was over.

**Comment by KTibow** [comment]
@jaredswarts55 here's your code modified a little (so it streams a certain amount):
```python3
import asyncio, time

from aiohttp import ClientSession
from simplipy import *


async def get_subscriptions(user_id: str, access_token: str, client: ClientSession):
    url = 'https://api.simplisafe.com/v1/users/{}/subscriptions'.format(user_id)
    async with client.get(url, headers={'Authorization': 'Bearer {}'.format(access_token)}) as resp:
        return await resp.json()

async def get_camera_feed(camera_id: str, access_token: str, client: ClientSession, capture_time: int):
    start_time = time.time()
    url = 'https://media.simplisafe.com/v1/{}/flv?x=472&audioEncoding=AAC'.format(camera_id)
    with open('testing.flv', 'wb') as file:
        async with client.get(url, headers={'Authorization': 'Bearer {}'.format(access_token)}) as resp:
            async for data in resp.content.iter_chunked(1024):
                file.write(data)
                if time.time() - capture_time > start_time:
                    print("Captured "+str(capture_time)+" seconds of video, exiting.")
                    break

async def main() -> None:
    """Create the aiohttp session and run."""
    async with ClientSession() as session:
        simplisafe = await API.login_via_credentials(
            input("What's your email: "), input("What's your password: "), session=session
        )

        systems = await simplisafe.get_systems();
        for s in systems:
            system = systems[s]

        subscriptionsBody = await get_subscriptions(simplisafe.user_id, simplisafe.access_token, session)
        cameraId = subscriptionsBody["subscriptions"][0]["location"]["system"]["cameras"][0]["uuid"]
        await get_camera_feed(cameraId, simplisafe.access_token, session, int(input("How many seconds?")))

asyncio.run(main())
```
Sorry if I forgot to put `await` somewhere, I don't know `asyncio` programming that well.

**Comment by jhujasonw** [comment]
This is pretty darned cool.  And just to add some inspiration here, I modified the script that @jaredswarts55 posted to output the text to stderr and the data to stdout.  I was able to pipe that to ffmpeg which would then connect to an ffserver to rebroadcast the stream in a meaninful way, and connect that stream into ZoneMinder.  This has the potential to give you the ability to stream and store your videos on your own machine, but sadly it still requires you to have a subscription for the camera.

One major drawback, the SimpliSafe API silently stops sending video after 5 minutes. (4:58 to be more exact).  There are potential ways around this issue, like having a small buffer to use while you re-setup the connection, but I haven't gotten to playing with that.

**Comment by uberlinuxguy** [comment]
This is pretty darned cool. And just to add some inspiration here, I modified the script that @jaredswarts55 posted to output the text to stderr and the data to stdout. I was able to pipe that to ffmpeg which would then connect to an ffserver to rebroadcast the stream in a meaninful way, and connect that stream into ZoneMinder. This has the potential to give you the ability to stream and store your videos on your own machine, but sadly it still requires you to have a subscription for the camera.

One major drawback, the SimpliSafe API silently stops sending video after 5 minutes. (4:58 to be more exact). There are potential ways around this issue, like having a small buffer to use while you re-setup the connection, but I haven't gotten to playing with that.  It does look like there has to be about a 30 second cool-down on the API before making another request.  I really hope I am missing something and that limit can be overcome.  @nikonratm Have you seen the 5 minute limit?

**Comment by KTibow** [comment]
Anyone want to try to analyze the network traffic when you try to stream for 11 minutes and see how it bypasses the 5 minute limit? (Make sure to keep the device awake.)

**Comment by uberlinuxguy** [comment]
OH! This may not be a simplisafe limit!  It appears that this could be a timeout on the aiohttp ClientSession object in python...
https://docs.aiohttp.org/en/stable/client_reference.html

**Comment by KTibow** [comment]
@uberlinuxguy Looks better if you add `?highlight=timeout` to the end:
https://docs.aiohttp.org/en/stable/client_reference.html?highlight=timeout

**Comment by uberlinuxguy** [comment]
@KTibow The docs say if you set any or all of the timeouts to 0 or none they are disabled.  I am currently running that test to see what happens... 4 mins in...

**Comment by uberlinuxguy** [comment]
First test with the timeouts set to zero ran until 6:20s.  I think my camera may be flaky or something.

I'll keep trying and post my modified code later once I work out the kinks.

**Comment by nikonratm** [comment]
@uberlinuxguy awesome work. The feed is definitely kinda temperamental about long-term streaming, we have the same issues but never liked too deeply into it. One thought would be to consider taking X second chunks to save to disk. Eg this snippet gives 10 second mp4s which can easily be joined back together:

```
ffmpeg -re -headers "Authorization: Bearer xxx=" -i "https://54.82.224.248/v1/xxx/flv?x=640&audioEncoding=AAC" -vcodec copy -pix_fmt yuv420p -r 20 -b:v 132k -bufsize 264k -maxrate 132k -f segment -segment_time 10 -segment_format_options movflags=+faststart -reset_timestamps 1 segment%d.mp4
```

**Comment by KTibow** [comment]
> ```bash
> ffmpeg -re -headers "Authorization: Bearer xxx=" -i "https://54.82.224.248/v1/xxx/flv?x=640&audioEncoding=AAC" -vcodec copy -pix_fmt yuv420p -r 20 -b:v 132k -bufsize 264k -maxrate 132k -f segment -segment_time 10 -segment_format_options movflags=+faststart -reset_timestamps 1 segment%d.mp4
> ```

Might be doable with [ffmpeg-python](https://github.com/kkroening/ffmpeg-python#ffmpeg-python-python-bindings-for-ffmpeg)... we could [concat the two files](https://stackoverflow.com/questions/7333232/how-to-concatenate-two-mp4-files-using-ffmpeg)...

**Comment by KTibow** [comment]
Any updates on this? It sounds doable.

**Comment by bggardner** [comment]
Like @shamoon, I've implemented camera integration in a PHP script at [bggardner/simplisafe-proxy](https://github.com/bggardner/simplisafe-proxy).  The whole script takes a very minimalist approach, especially during authentication, as I only pass what seems to be necessary, instead of fully mimicking the WebApp.  Feel free to steal ideas or ask questions (though it's been a while since I've looked at it).

**Comment by KTibow** [comment]
Go away bot, it's not solved.

**Comment by bachya** [comment]
@KTibow This is a valid use case of the bot: it's been over 30 days since this issue was last touched. Let's add the `help wanted` label.

**Comment by jhujasonw** [comment]
I had a stream that was working, but abandoned the idea for a few reasons:

- It required a multi-step process to get it into something useful like motion-eye.  (ie, python script to read from simplisafe, ffmpeg to reformat and send to an ffserver for RTSP so motion-eye could read it)
- The stream was flaky and I could not come up with a good way to buffer it as it would go out for >30 seconds.  This could have been my wifi, which is highly possible, but it was so unstable it was unusable.
- When the camera was streaming it was not available in the simplisafe app.  Attempting to access it would not work and would actually hard reset the camera.
- Streaming the camera also seemed to turn off simplisafe's motion detection.

If someone is still interested in working on this, I may be able to dig up the python I was using, but it was a small modification of what is already posted here to stream endlessly to stdout which was redirected to ffmpeg

**Comment by bachya** [comment]
Quick comment: #182 added initial support for this, but the comments here (@jhujasonw's more recently) are still worth consideration before we can consider this fully closed. So, going to leave this open for further thought and effort.

**Comment by veedubb65** [comment]
I have a simplisafe doorbell camera. Is there anything I can do to assist with getting the cameras (especially the doorbell camera) better integrated?

**Comment by bachya** [comment]
@veedubb65 Read all the above and see where you can help.

**Comment by jarodwilson** [comment]
Hey @bachya, thanks for your help getting my "refuse to pay for any level of service" simplisafe setup working, and as we discussed, I've got the doorbell with camera that I'd be more than happy to hack on.

**Comment by nprez83** [comment]
Hi everyone, wondering if there's been any progress on accessing the Simplisafe Cam stream from HA. I'd love to be able to do that. Thanks @bachya for all the hard work you've put into this awesome integration, definitely one of the most important (if not the most) integrations I have.

**Comment by peebles** [comment]
For the outdoor camera, in case you haven't seen the timeline response, looks like this:

**GET https://api.simplisafe.com/v1/subscriptions/6720379/events?numEvents=50**

```json
{
    "numEvents": 50,
    "lastEventTimestamp": 1697547602,
    "events": [
        {
            "eventId": 35952684503,
            "eventTimestamp": 1697598147,
            "eventCid": 1170,
            "zoneCid": "1",
            "sensorType": 17,
            "sensorSerial": "f11b6abd",
            "account": "00499FF8",
            "userId": 5913258,
            "sid": 6720379,
            "info": "Camera Detected Motion",
            "pinName": "",
            "sensorName": "Back Yard",
            "messageSubject": "Camera Detected Motion",
            "messageBody": "Camera Detected Motion on 10/17/2023 at 8:02 PM",
            "eventType": "activityCam",
            "timezone": 3,
            "locationOffset": -420,
            "videoStartedBy": "6172311af9da430ab2e11c59f11b6abd",
            "video": {
                "6172311af9da430ab2e11c59f11b6abd": {
                    "clipId": 24021857466,
                    "preroll": 1,
                    "postroll": 12,
                    "cameraName": "Back Yard",
                    "cameraModel": "SSOBCM4",
                    "aspectRatio": "16:9",
                    "duration": 13,
                    "recordingType": "KVS",
                    "account": "611485993050",
                    "region": "us-east-1",
                    "_links": {
                        "_self": {
                            "href": "https://chronicle.us-east-1.prd.cam.simplisafe.com/v1/recordings/24021857466",
                            "method": "GET"
                        },
                        "preview/mjpg": {
                            "href": "https://remix.us-east-1.prd.cam.simplisafe.com/v1/preview/6172311af9da430ab2e11c59f11b6abd/6720379/time/1697598146/1697598159?account=611485993050&region=us-east-1{&fps,width}",
                            "method": "GET",
                            "templated": true
                        },
                        "snapshot/mjpg": {
                            "href": "https://remix.us-east-1.prd.cam.simplisafe.com/v1/preview/6172311af9da430ab2e11c59f11b6abd/6720379/time/1697598146/1697598150?account=611485993050&region=us-east-1{&fps,width}",
                            "method": "GET",
                            "templated": true
                        },
                        "snapshot/jpg": {
                            "href": "https://remix.us-east-1.prd.cam.simplisafe.com/v1/snapshot/6172311af9da430ab2e11c59f11b6abd/6720379/time/1697598146?account=611485993050&region=us-east-1{&width}",
                            "Method": "GET",
                            "templated": true
                        },
                        "download/mp4": {
                            "href": "https://remix.us-east-1.prd.cam.simplisafe.com/v1/download/6172311af9da430ab2e11c59f11b6abd/6720379/time/1697598146/1697598159?account=611485993050&region=us-east-1",
                            "method": "GET"
                        },
                        "share": {
                            "href": "https://remix.us-east-1.prd.cam.simplisafe.com/v2/share/6172311af9da430ab2e11c59f11b6abd/6720379/time/1697598146/1697598159?account=611485993050&region=us-east-1",
                            "method": "POST"
                        },
                        "playback/dash": {
                            "href": "https://mediator.prd.cam.simplisafe.com/v1/recording/6172311af9da430ab2e11c59f11b6abd/6720379/time/1697598146/1697598159/dash?account=611485993050&region=us-east-1",
                            "method": "GET"
                        },
                        "playback/hls": {
                            "href": "https://mediator.prd.cam.simplisafe.com/v1/recording/6172311af9da430ab2e11c59f11b6abd/6720379/time/1697598146/1697598159/hls?account=611485993050&region=us-east-1",
                            "method": "GET"
                        }
                    }
                }
            }
        },
...
```

I used curl on the snapshot/jpg url (with the Authorization: Bearer token) and got a good snap from the camera.

**Comment by mginz83** [comment]
> I don't really know what I am doing in python but I was able to stream the flv to a file and watch it with vlc using this code:
> 
> ```python
> import asyncio
> 
> from aiohttp import ClientSession
> from simplipy import *
> 
> 
> async def get_subscriptions(user_id: str, access_token: str, client: ClientSession):
>     url = 'https://api.simplisafe.com/v1/users/{}/subscriptions'.format(user_id)
>     print("Url To Send Call To: {}".format(url))
>     async with client.get(url, headers={'Authorization': 'Bearer {}'.format(access_token)}) as resp:
>         return await resp.json()
> 
> async def get_camera_feed(camera_id: str, access_token: str, client: ClientSession):
>     url = 'https://media.simplisafe.com/v1/{}/flv?x=472&audioEncoding=AAC'.format(camera_id)
>     print("Url To Send Call To: {}".format(url))
>     with open('testing.flv', 'wb') as file:
>         async with client.get(url, headers={'Authorization': 'Bearer {}'.format(access_token)}) as resp:
>             async for data in resp.content.iter_chunked(1024):
>                 file.write(data)
> 
> async def main() -> None:
>     """Create the aiohttp session and run."""
>     async with ClientSession() as session:
>         simplisafe = await API.login_via_credentials(
>             "YOUR_EMAIL_HERE", "YOUR_PASSWORD_HERE", session=session
>         )
> 
>         systems = await simplisafe.get_systems();
>         for s in systems:
>             system = systems[s]
> 
>         subscriptionsBody = await get_subscriptions(simplisafe.user_id, simplisafe.access_token, session)
>         cameraId = subscriptionsBody["subscriptions"][0]["location"]["system"]["cameras"][0]["uuid"]
>         print("Camera Id: {}".format(cameraId))
>         print("Subscriptions: {}".format(subscriptionsBody))
>         await get_camera_feed(cameraId, simplisafe.access_token, session)
> 
> 
> asyncio.run(main())
> ```

how do we look at the feed? can we call this API from, lets say javascript or react?

**Comment by KTibow** [comment]
> I took this URL into a browser and it didnt work

The code should be pretty easily portable if you consider things like sending authorization

**Comment by Zippyduda** [comment]
I'd love to see this working so I could view the SimpliSafe cameras in Home Assistant (running HAOS on a Raspberry Pi 4) like I do with Hikvision currently. I've got SimpliSafe on the Pro Premium plan (UK) in case anyone needs help testing this to get it working. I'm no coder sadly.

**Comment by peebles** [comment]
For home assistant support, its not in this repo that changes would need to be made.  https://github.com/home-assistant/core.git, under homeassistant/components/simplisafe/, the code in HA that uses this repo.

**Comment by peebles** [comment]
Locally, I have made modifications to home-assistant/core/homeassistant/components/simplisafe for the OUTDOOR_CAMERA as described below, and I wonder if this is a good approach.  

I treat the OUTDOOR_CAMERA as a binary_sensor (motion).  This new sensor listens for EVENT_CAMERA_MOTION_DETECTED coming from the Simplisafe websocket.  When this event occurs it fetches the "event list" (or timeline) from Simplisafe server and looks for events targeted at the right device serial number, occurring at the same timestamp as the websocket event.  These raw SS timeline events contain the media urls for the motion event (see my comment above for the format).  I download the snapshot/jpg and download/mp4 files and store them under /config/www/simplisafe/<camera-name>/[latest_snapshot.jpg, latest_clip.mp4] and ./clips/YYYYMMDDHHmmss.mp4.  I use the HACS gallery card to browse the clips.  I use the latest_snaphot.jpg in notification automations.

It does what I want for the OUTDOOR_CAMERA.  Its not a camera, in that there is no way to stream the current video ... there does not seem to be a good way of doing this for these camera types.

**Comment by pcmoore** [comment]
It may not be a full-fledged video feed, but it is much better than what we currently have (nothing).  I'm not sure what the HA maintainers think, but I would be excited to have this level of camera support in the SimpliSafe integration.

Does it work for the doorbell too, or just the outdoor camera?

**Comment by mginz83** [comment]
Can you share your implementation? Ive tried many ways but cant get this one to work for me?

**Comment by peebles** [comment]
I'll go through the developer checklist and do a pull request, maybe today.

**Comment by peebles** [comment]
@mginz83 So you know how to run an integration with changes locally in a running HA?  If you do, you'll find my changes in https://github.com/peebles/core/tree/simplisafe-outdoor-motion-capture/homeassistant/components/simplisafe, specifically in `binary_sensor.py`.  Comments in the code will give you some idea how to use it.  

I have submitted a PR, but it might take a while to be approved and released.

**Comment by pcmoore** [comment]
@peebles can you link to the PR here?  It would be nice to see a record of all the various attempts in this issue.

**Comment by KTibow** [comment]
Link: https://github.com/home-assistant/core/pull/106008

**Comment by ccocks** [comment]
I modified @jaredswarts55 solution to use an Auth Code/Code Verifier:
```python

import asyncio

from aiohttp import ClientSession
from simplipy import *


async def get_subscriptions(user_id: str, access_token: str, client: ClientSession):
    url = 'https://api.simplisafe.com/v1/users/{}/subscriptions'.format(user_id)
    print("Url To Send Call To: {}".format(url))
    async with client.get(url, headers={'Authorization': 'Bearer {}'.format(access_token)}) as resp:
        return await resp.json()

async def get_camera_feed(camera_id: str, access_token: str, client: ClientSession):
    url = 'https://media.simplisafe.com/v1/{}/flv?x=472&audioEncoding=AAC'.format(camera_id)
    print("Url To Send Call To: {}".format(url))
    with open('testing.flv', 'wb') as file:
        async with client.get(url, headers={'Authorization': 'Bearer {}'.format(access_token)}) as resp:
            async for data in resp.content.iter_chunked(1024):
                file.write(data)

async def main() -> None:
    """Create the aiohttp session and run."""
    async with ClientSession() as session:
        simplisafe= await API.async_from_auth(
            "BFHQc62Rfb6HdUhbcATWVNffUlNxgWfnSEWank1IldUyP",
            "GWvuu5a22As6EVvGuoLMFkFZqangk2GSqkYSeDRhl52aTB96yg",
            session=session,
        )

        systems = await simplisafe.async_get_systems()
        for s in systems:
            system = systems[s]

        subscriptionsBody = await get_subscriptions(simplisafe.user_id, simplisafe.access_token, session)
        #print("Camera: " + dir(subscriptionsBody["subscriptions"][0]["location"]["system"]["cameras"][0]))
        
        cameraId = subscriptionsBody["subscriptions"][0]["location"]["system"]["cameras"][0]["uuid"]
        print("Camera Id: {}".format(cameraId))
        print("Subscriptions: {}".format(subscriptionsBody))
        await get_camera_feed(cameraId, simplisafe.access_token, session)


asyncio.run(main())
``` 

But it streamed to a flv file with 0 bytes. Any idea why?

**Comment by KTibow** [comment]
Perhaps the request failed. Check the status of `resp`.

**Comment by psbankar** [comment]
Just wanted to check if this now supports simplicam on free subscription. I could not access the camera from my end

**Comment by athanasio** [comment]
Hello, it might be tangent to the rest of the discussion but I thought it's better to ask here: is it possible to have some basic camera integration without the streams per-se? e.g. ability to have motion detection and motion notifications as binary sensors, sensitivity as entity etc. It would be useful to be able to create automations that enable/disable features, even if we can't view the streams on HA.

**Comment by rlippmann** [comment]
@peebles Did you resubmit the PR?

@athanasio That's already possible via HA event triggers (although probably will take a little debugging to figure out the payload that the websocket will send).  But, if cameras are implemented, it probably would make more sense to implement it internally as binary sensors on a camera device.

**Comment by RevillWeb** [comment]
I have several cameras, let me know if I can help in any way.

**Comment by zouhairm** [comment]
I resubmitted PR here: https://github.com/home-assistant/core/pull/139663

tested it locally using 2025.2.5 and seeing camera entities

**Comment by micchickenburger** [comment]
Hey @zouhairm, wondering if you have a chance soon to address the PR requested changes.  Would love to see this support soon :)

**Comment by zouhairm** [comment]
Hey sorry I missed the github notifications for this.
What is the PR requesting, documentation sounds like? looks that's a new thing since this PR was first originally created: I'll try to understand what I need to do, but if you know where I can read up on how to put documentation together I can try to work on that

**Comment by alerum68** [comment]
Looks like it was closed and locked. Will have to resubmit a new PR.

**Comment by gilliginsisland** [comment]
I know this is an old thread, but I have gotten streams to work using go2rtc.

The SimpliSafe webrtc stream in the browser is using Amazon Kinesis which go2rtc supports natively.

I use this config for go2rtc

```yaml
streams:
  simplirtc_dining_room:
    - echo:/opt/uv/bin/simplirtc --token=/etc/go2rtc/simplirtc.token stream --location= 1234567 --camera=2e62e933cd05957f502797cf95d6fa49
    - ffmpeg:simplirtc_dining_room#audio=opus/16000

  simplirtc_basement:
    - echo:/opt/uv/bin/simplirtc --token=/etc/go2rtc/simplirtc.token stream --location= 1234567 --camera=c725b9ecaa40be2e1c748d198c0e70ba
    - ffmpeg:simplirtc_basement#audio=opus/16000
```

Where `/opt/uv/bin/simplirtc` is a small python wrapper around this library that extracts the kenisis parameters and passes them to go2rtc. https://github.com/gilliginsisland/simplirtc


You generate a token by running `/opt/uv/bin/simplirtc --token <path-to-token-file> authenticate` and go through the interactive prompt. 

Then use `/opt/uv/bin/simplirtc --token <path-to-token-file> cameras` to get a list of location ids and camera ids. 

You will get an output like 
```json
{
  "My Home (1234567)": [
    "Dining Room (2e62e933cd05957f502797cf95d6fa49)",
    "Basement (c725b9ecaa40be2e1c748d198c0e70ba)"
  ]
}
```

go2rtc will convert the stream to whichever format your using, whether its webrtc in the browser, or rtsp. you can then add the rtsp stream to anything you want, homeassistant, frigate, etc. go2rtc also has HomeKit support, and I've been using this to add my simplisafe cameras to HomeKit for a few months now and its been working great.

**Comment by pcmoore** [comment]
Thanks @gilliginsisland, this looks like a big step forward!  As someone who is very interested in integrating SimpliSafe cameras with my HA install, a few questions:

* Reading through the simplirtc README and your description above, it isn't clear to me how battery powered camera are handled.  Does this approach run the camera continuously, or is the camera stream still subject to the SimpliSafe configuration, e.g. motion activated?
* If the camera stream remains intermittent and triggered according to the SimpliSafe configuration, how do applications like HA, Frigate, etc. handle the starting/stopping of the video stream?
* Is there a way to simply capture an image from camera upon request to save on battery life?
* Previous attempts had run into problems when interacting with the existing SimpliSafe apps, tools, etc., does this approach negatively impact the Simplisafe app or service?

Once again, thanks for your work on this!  I'm really hoping we'll have a proper addon/integration for SimpliSafe cameras in the near future!

**Comment by archerne** [comment]
@gilliginsisland This doesn't appear to work if you don't have a simplisafe subscription. Or something else is going on. I can authenticate, however when i try and retrieve cameras it comes back as `{}`

![Image](https://github.com/user-attachments/assets/7d93ac86-a8a0-41d9-a32b-a170bce97413)

**Comment by gilliginsisland** [comment]
@pcmoore 

TLDR; The stream is only running on-demand

My script doesn't deal with motion detection or anything, it only deals with on-demand watching of the stream. 

When you try to watch the stream from go2rtc, it will execute my script to retrieve the stream details, the script just makes the api calls to extract those details and prints them to stdout and exits right away.

Go2rtc handles the kenisis webrtc connection and lifecycle. One of the things that go2rtc does, is that it will make a single connection to any stream it proxies and it will multiplex that to any clients connected to it. The stream is done on-demand so that when the last client disconnects from the stream, after a small timeout (not sure how long it is by default, but I think less than a minute) it will kill the connection.

From the SimpliSafe api side, nothing is required to kill the connection, as soon as the kenisis client is done, the camera stops.

In terms of getting a snapshot, I think that go2rtc has an api endpoint for fetching a snapshot. Internally it does this by activating the stream, taking the snapshot, and returning the image.

I haven't had the patience to look at the homeassitant integration, but with the recent support in HA for cameras returning a webrtc stream, it should be possible to have a python kenisis client and add this stream to the SimpliSafe integration directly, skipping the go2rtc part. I haven't needed it because I use go2rtc for exposing the cameras to HomeKit anyways, so this was enough for me.

**Comment by gilliginsisland** [comment]
@archerne I have a subscription and haven't tested without one. The way I got the stream information was by viewing the cameras in my browser and then checking the network logs. Can you confirm if you are able to view the cameras in your browser when using an unpaid account?

**Comment by archerne** [comment]
@gilliginsisland yes I can see the camera in my browser in an unpaid account

![Image](https://github.com/user-attachments/assets/6cce5783-8492-4609-a57f-8e45a98c0c2b)

**Comment by gilliginsisland** [comment]
@pcmoore if the go2rtc solution worked for you, I ported the kinesis stream logic to python and made a homeassistant integration for use with hacs that adds WebRTC support for simplisafe to homeassistant.

Basically you add this repo https://github.com/gilliginsisland/hass-simplirtc to HACS and then install the Simplirtc integration. Add `simplirtc:` to your `configuration.yaml` and it will automatically add any cameras with webrtc live view support from the native homeassistant simplisafe integration, with no reliance on go2rtc.

**Comment by gilliginsisland** [comment]
@bachya I am too lazy to add the correct PR to this repo, but feel free to copy the code from https://github.com/gilliginsisland/hass-simplirtc/blob/main/custom_components/simplirtc/camera.py#L176 for getting the kinesis details, and https://github.com/gilliginsisland/hass-simplirtc/blob/main/custom_components/simplirtc/kinesis.py#L74 for how to handle the signaling.

**Comment by sswires** [comment]
@gilliginsisland I tried your solution and the cameras show up as devices in HA, but the live stream doesn't work. When viewing the log, I see the following:

```
Logger: simplipy
Source: custom_components/simplirtc/camera.py:178
integration: Simplisafe WebRTC Cameras ([documentation](https://github.com/gilliginsisland/hass-simplirtc), [issues](https://github.com/gilliginsisland/hass-simplirtc/issues))
First occurred: 22:08:53 (2 occurrences)
Last logged: 22:09:01
Giving up _async_api_request(...) after 1 tries (aiohttp.client_exceptions.ClientResponseError: 404, message='Not Found', url='https://app-hub.prd.aser.simplisafe.com/v2/cameras/<redacted camera ID>/<redacted user ID>/live-view')
```

When I use the web app, the live stream of my doorbell camera, it comes from a URL like the following `https://media.simplisafe.com/v1/<camera ID>/flv?x=590&audioEncoding=AAC`. However, I noticed my battery powered camera does use the expected URL scheme but that doesn't work either. However, I noticed the response object differs from how your implementation is expecting it, and it seems to be using something called "livekit". This gives a response that looks like the following:

```
{
  "liveKitDetails": {
    "liveKitURL": "wss://livestream.services.simplisafe.com:7880",
    "userToken": "<redacted>"
  },
  "cameraStatus": "online"
}
```

This Websocket server seems to be using [Livekit](https://github.com/livekit/python-sdks) as the protocol looks like Protobuf, and that's how it eventually gets the stream.

If you want to discuss this on your own repo then I can potentially share more details.

**Comment by gilliginsisland** [comment]
@sswires 

I didn't realize that they use different methods for webrtc. I will have to add a filter to check for `provider == kinesis` 

Based on the port 7880 and the use of protobuf, I think it might actually be this go library 
https://github.com/livekit/livekit
https://docs.livekit.io/home/self-hosting/ports-firewall/

I don't have a camera that uses that to test, but am open to any PRs that add support.

**Comment by AverageHelper** [comment]
> Basically you add this repo https://github.com/gilliginsisland/hass-simplirtc to HACS and then install the Simplirtc integration. Add simplirtc: to your configuration.yaml and it will automatically add any cameras with webrtc live view support from the native homeassistant simplisafe integration, with no reliance on go2rtc.

I'm just here to confirm this does work. I don't have HACS, but I installed hass-simplirtc manually, added `simplirtc:` to the config, and now camera entities are available along with the usual door sensors and alarm controls, etc.! 🎉 

EDIT: So far all but one of the cameras work (I only have a handful); I'll do some testing later to see why that might be.

**Comment by gilliginsisland** [comment]
When I originally wrote this my simplisafe cameras were using kinesis, but they have now switched to livekit for me as well.

I updated my hass plugin to support livekit, and I tried to maintain backwards compatibility with kinesis but since my cameras have started using livekit i have no way to actually test that.
@sswires @AverageHelper Please test my newest update v1.2 and let me know if it works.

The livekit method requires a proxy that I wrote in golang which uses the livekit sdk and is included in the hass integration. The integration will fetch the lastest binary from my github and will manage it. The downside is that instead of the browser direct connecting to livekit server, instead the homeassistant server will be proxying the stream.

The reason a proxy is required is because livekit does not work with the way homeassistant expects webrtc to work. homeassistant gives an integration a webrtc offer, and expects a single answer with both audio and video tracks in the same answer. Livekit works the opposite and has the video server send the offer and wants an answer. it also uses separate peer connections and requires advanced signalling, so either SDP munging is required, or a custom plugin using their sdk can expose an rtsp stream.

I am currently working on doing raw webrtc and seeing if i can munge the sdps to be compatible, but I am unsure if it will work. If it does work i will be able to eliminate the proxy.

**Comment by sswires** [comment]
Hi @gilliginsisland 

I updated to 1.2, but I'm not seeing my doorbell camera anymore. It shows as a device but has no entities. 

Thanks

**Comment by zjleblanc1** [comment]
I've been trying to get this configured for a few weeks. Before the update, the cameras were listed but the video wouldn't load. After the update, the entities now say "Unavailable" and "This entity is no longer being provided by the simplirtc integration." 

The logs show "Camera 'Front Door' has unknown webrtc provider 'simplisafe'".

**Comment by repairman29** [comment]
🎯 **PING RECEIVED!**

**@repairman29** has claimed this bounty via [Echeo](https://echeo.io).

**Match Score:** 85%  
**Status:** Ready to start working!

[View conversation in Echeo](https://echeo.io/messages) to coordinate.

**Comment by repairman29** [comment]
Hey there! This looks like something I could help with. I'd be happy to take a look and see what we can do.

*(Found via [Echeo](https://echeo.io))*

**Comment by derreck** [comment]
How do you set this up? I've added the repo in hacs and now what do i do? i've read add simplirtc to config.yaml but that doesn't do anything but throw and error. Could someone explain the steps to getting this configured please?
